pipeline/nissl_to_ccfv3.py

```diff
@@ -0,0 +1,223 @@
+        return 1
+    volume = load_volume(path, normalize=False)
+    return volume.astype(np.float32)
+
+
+def slice_registration(
+    fixed: np.array, moving: np.array, nissl_slice: np.array | None = None
+) -> tuple[np.array, np.array | None]:
+    """Compute registration transform between a couple of slices.
+
+    Parameters
+    ----------
+    fixed
+        Fixed slice of shape (h, w) of type annotation.
+    moving
+        Moving slice of shape (h, w) of type annotation.
+    nissl_slice
+        If needed, intensity slice of shape (h, w)  in the same coordinate
+        system as moving.
+
+    Returns
+    -------
+    warped : np.array
+        Warped slice of type annotation.
+    nissl_slice : np.array | None
+        If a nissl slice is specified, the given slice is warped with the
+        same transformation.
+    """
+    nii_data = register(fixed, moving, is_atlas=True)
+    warped = transform(moving, nii_data, interpolator="genericLabel")
+
+    if nissl_slice is not None:
+        nissl_slice = transform(nissl_slice, nii_data)
+
+    return warped, nissl_slice
+
+
+def registration(
+    reference_volume: np.array, moving_volume: np.array, nissl_volume: np.array
+) -> tuple[np.array, np.array]:
+    """Compute registration between a moving volume and reference one.
+
+    The registration is computed in two steps:
+    1. First between one slice of the moving volume and the neighbouring slice
+    (To reduce the jaggedness)
+    2. Once the first registration is done, the transformation between the
+    resulting slice and the corresponding slice of the reference volume is
+    applied. (To make moving and reference more similar)
+    The registration is computed from the center of the volume to the two
+    extremities.
+
+    Parameters
+    ----------
+    reference_volume
+        Reference volume.
+    moving_volume
+        Moving volume of the same shape as the moving_volume.
+    nissl_volume
+        Nissl volume to register. It has to have same shape as reference_volume
+        and moving_volume and be from the same coordinate system as the moving
+        volume.
+
+    Returns
+    -------
+    warped_volume
+        Moving volume after the registration transformation.
+    nissl_warped
+        Nissl volume once the registration transformation are applied.
+    """
+    warped_volume = np.zeros_like(moving_volume)
+    nissl_warped = np.zeros_like(nissl_volume)
+    total_n_iterations = moving_volume.shape[0]
+    n_iterations = int(total_n_iterations / 2)
+
+    fixed = moving_volume[n_iterations, :, :]
+    for i in range(n_iterations):  # 0 -> 263
+        index = n_iterations - (i + 1)
+        warped = moving_volume[index]
+        fixed_ref = reference_volume[index]
+        nissl_slice = nissl_volume[index]
+
+        warped, nissl_slice = slice_registration(fixed, warped, nissl_slice)
+        warped, nissl_slice = slice_registration(fixed_ref, warped, nissl_slice)
```

Comment:
Is this really helping to prevent the jaggedness from occurring (empirically)?

My intuition is that the second registration should give us the same result no matter how we modified/preregistered the `moving` image.

Reply:
We dicussed this in person